drift-transport-netty/src/main/java/io/airlift/drift/transport/netty/server/ThriftServerHandler.java

```diff
@@ -0,0 +1,326 @@
+import io.airlift.drift.transport.server.ServerInvokeRequest;
+import io.airlift.drift.transport.server.ServerMethodInvoker;
+import io.airlift.log.Logger;
+import io.airlift.units.Duration;
+import io.netty.buffer.ByteBuf;
+import io.netty.channel.ChannelDuplexHandler;
+import io.netty.channel.ChannelHandlerContext;
+
+import java.lang.reflect.Type;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.OptionalInt;
+import java.util.concurrent.ScheduledExecutorService;
+
+import static io.airlift.drift.TApplicationException.Type.INTERNAL_ERROR;
+import static io.airlift.drift.TApplicationException.Type.INVALID_MESSAGE_TYPE;
+import static io.airlift.drift.TApplicationException.Type.UNKNOWN_METHOD;
+import static io.airlift.drift.protocol.TMessageType.EXCEPTION;
+import static io.airlift.drift.protocol.TMessageType.REPLY;
+import static java.util.Objects.requireNonNull;
+import static java.util.concurrent.TimeUnit.MILLISECONDS;
+
+public class ThriftServerHandler
+        extends ChannelDuplexHandler
+{
+    private static final Logger log = Logger.get(ThriftServerHandler.class);
+
+    private final ServerMethodInvoker methodInvoker;
+    private final ScheduledExecutorService timeoutExecutor;
+    private final Duration requestTimeout;
+
+    public ThriftServerHandler(ServerMethodInvoker methodInvoker, Duration requestTimeout, ScheduledExecutorService timeoutExecutor)
+    {
+        this.methodInvoker = requireNonNull(methodInvoker, "methodInvoker is null");
+        this.requestTimeout = requireNonNull(requestTimeout, "requestTimeout is null");
+        this.timeoutExecutor = requireNonNull(timeoutExecutor, "timeoutExecutor is null");
+    }
+
+    @Override
+    public void channelRead(ChannelHandlerContext context, Object message)
+            throws Exception
+    {
+        if (message instanceof ThriftFrame) {
+            messageReceived(context, (ThriftFrame) message);
+            return;
+        }
+        context.fireChannelRead(message);
+    }
+
+    private void messageReceived(ChannelHandlerContext context, ThriftFrame frame)
+    {
+        try {
+            ListenableFuture<ThriftFrame> response = decodeMessage(
+                    context,
+                    frame.getProtocolFactory(),
+                    frame.getMessage(),
+                    frame.getHeaders(),
+                    frame.isSupportOutOfOrderResponse());
+            Futures.addCallback(response, new FutureCallback<ThriftFrame>()
+            {
+                @Override
+                public void onSuccess(ThriftFrame result)
+                {
+                    context.writeAndFlush(result);
+                }
+
+                @Override
+                public void onFailure(Throwable t)
+                {
+                    context.disconnect();
+                }
+            });
+        }
+        catch (Exception e) {
```

Comment:
Does this happen only for bugs? Or can we get here if the client sends garbage data?

Reply:
Should this also close the connection on Throwable?